### PR TITLE
[connector/count] Nested attributes as dimensions

### PR DESCRIPTION
**Pull request by greatestusername**
**Description:**
Adding a feature: Allow the use of nested attributes as dimensions (as far as I know this is only common in logs) in count connector.
This is essentially ripped from the loki translator pkg.

**Link to tracking Issue:** [19930](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/19930) 

**Testing:** Locally and works. 

Please describe best approach to testing this case with given testing rig? Tried to work out the best way to write tests and appropriate test data but ended up with more questions than solutions.

Tried creating a `nested_attribute.json` with testing data and an appropriate config along with some test data in `input.json` but feels like I'm missing some vital information. Do I need to create a new resource in `input.json` and each other json file in addition to making a new json file for testing nested attributes? A little lost as the best path forward here. 

**Documentation:** No new documentation. This new work just allows parsing nested attribute names.

### CODE REVIEW

**Review by djaglowski — COMMENTED** [review]
Thanks for putting this together @greatestusername.

I think I misunderstood what you were suggesting. It looks like the idea here is to essentially search for a specific attribute key regardless of where it is found.

Typically, when we work with nested attributes they are understood in context, so if you have
```json
{
  "foo": {
    "hello": "world"
  },
  "bar": {
    "hello": "world"
  },
  "hello": "world"
}
```
then we understand these as three separate attributes, `foo.hello`, `bar.hello`, and `hello` as opposed to three instances of `hello`.

That said, I think your use case could be supported, but instead of just automatically searching nested attributes, I think we should have a flag in the logs config that enables this functionality. Possibly, `any_depth: true`, but I'm open to ideas.

As far as testing goes, here's what I would suggest. 
1. Try to add a complex attribute to the existing `testdata/logs/input.json`. This should contain an attribute key that you want to count. This may work:
```json
{
    "key": "foo",
    "value": {
        "kvlistValue": {
            "values": [
                {
                    "key": "hello",
                    "value": {
                        "stringValue": "world"
                    }
                },
                {
                    "key": "bar",
                    "value": {
                        "kvlistValue": {
                            "values": [
                                {
                                    "key": "hello",
                                    "value": {
                                        "stringValue": "world"
                                    }
                                }
                            ]
                        }
                    }
                }
            ]
        }
    }
}
```
2. Run the current test cases to make sure it hasn't broken anything. Ideally the addition of this complex attribute did not change any other counts, since it would normally be ignored.
3. Add a new test case to `TestLogsToMetrics`, and configure it to look for "hello" (if you use the above example).

**Comment by greatestusername** [comment]
Thank you for your reply!

So the more I looked at this it looks like specifically an issue with parsing logs? Because I was not able to get say `foo.hello` value from this type of log data which was why I made the original issue to search for a specific nested attribute key like `foo.hello`:
```
{
    "key": "foo",
    "value": {
        "kvlistValue": {
            "values": [
                {
                    "key": "hello",
                    "value": {
                        "stringValue": "world"
                    }
                },
                {
                    "key": "bar",
                    "value": {
                        "kvlistValue": {
                            "values": [
                                {
                                    "key": "hello",
                                    "value": {
                                        "stringValue": "world"
                                    }
                                }
                            ]
                        }
                    }
                }
            ]
        }
    }
}
```
Am I missing something and that is expected to return `foo.hello` currently? Or maybe I am misunderstanding?

Appreciate the test pointers. I'll try again with that knowledge in mind.
Thanks!

**Comment by djaglowski** [comment]
> Am I missing something and that is expected to return foo.hello currently?

Apologies, I should have been clear that I was referring to `foo.hello` in an abstract way to indicate that `hello` is not just `hello` in this case, but that it is "`hello` that is a sub-attribute of `foo`".

We do have some syntax in other parts of the collector which can refer to nested attributes (e.g. `foo["hello"]`) but it's not clear to me that we could make much use of that here. The challenge is that we can't access a nested attribute directly [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/95aada5f3ceed5a879f69cd89b658edd85fd1e92/connector/countconnector/counter.go#L54). We'll have to drill down and find it recursively. We can probably maintain counts for nested attributes in a flat map but I think the end result needs to reflect the actual structure. e.g.

```json
{
  "hello": 3,
  "foo": {
    "hello": 2
  },
  "bar": {
    "hello": 4
  }
}
```

**Comment by greatestusername** [comment]
Ok gotcha. Yeah I'm definitely noticing the map situation when testing.

So thoughts on path forward? Current code totally works on logs and gets just the intended key. It can't traverse into slices but warns when that is attempted.

Thoughts? Should I be trying to produce a flat map or some other structure?

Again appreciate your help and eyes on this!

**Comment by djaglowski** [comment]
I think the proper thing to do is to keep track of counts separately and produce a structured result that reflects counts. However this is pretty complicated.

I think it would be ok to use what you have as long as it's disabled by default. See my comment about `any_depth: true` [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20421#pullrequestreview-1361436428).

So basically, the structure of the results would be unchanged. The only thing that would differ when `any_depth` is `true`, would be that we search into nested attributes and accumulate into the same counter.

**Comment by greatestusername** [comment]
That makes sense to me. I'll get to adding an any depth setting and see about getting it all tested up.

Thanks!!

**Comment by djaglowski** [comment]
I think just supporting logs makes sense, for now at least.

**Comment by greatestusername-splunk** [comment]
Sounds good!

**Comment by djaglowski** [comment]
Thanks for your patience on this. I think we're making progress.

> This function is only called by resource configs that are not logs.
>
> Was looking for a better way to restrict this functionality/restriction and test it but what I've got seemed the most straight forward. I'd prefer to have the check be part of the validateAttributes() function but there isn't currently a way to pass along what type of resource the config is defining (spans, spanevents, datapoints, etc).

I'm wondering now if we should just allow the setting to apply for any AttributeConfig. Initially I was thinking we should only support it for logs because that might be easier, but if I'm looking at it correctly, it seems it's actually more complicated to enforce the restriction. What do you think?

Another thought, I can imagine that once this is available, someone will ask for a way to limit the depth. We could handle this from the start by changing this to `max_depth` with a default of 1. WDYT?

**Comment by greatestusername** [comment]
I'm fine making it allowable for any AttributeConfig. Having the ability to enable config selectively for different types of attributes may be useful at some point but for now maybe unwarranted.

Regarding max depth I wonder about the utility of this versus the confusion it might create. For the example below it seems like the depth is pre-defined (`foo.hello` would have a depth of 1 and `foo.bar.hello` would have a depth of 2):

```{
    "key": "foo",
    "value": {
        "kvlistValue": {
            "values": [
                {
                    "key": "hello",
                    "value": {
                        "stringValue": "world"
                    }
                },
                {
                    "key": "bar",
                    "value": {
                        "kvlistValue": {
                            "values": [
                                {
                                    "key": "hello",
                                    "value": {
                                        "stringValue": "world"
                                    }
                                }
                            ]
                        }
                    }
                }
            ]
        }
    }
}
```
What happens when `max_level` is set to `1` for `foo.bar.hello` in these cases? I may also be totally misunderstanding the intent of adding `max_level` though. Let me know what you think?
 
As always appreciate your eyes and review! Thank you!

**Comment by djaglowski** [comment]
Any time recursion is involved, I think we need to have a reliable way to limit the depth. Otherwise we have potential concerns about performance and stability. 

My intention with suggesting `max_depth` with a default of 1, is that it would behave the same as the connector does now. So `hello` would be counted, but `foo.hello` and `foo.bar.hello` would not. If we want to capture `foo.hello`, we could increase `max_depth` to 2, or we could capture all of these with any setting greater than 2.

Is the confusion based on whether the depth is zero-indexed vs one-indexed?

**Comment by greatestusername** [comment]
Yeah totally understand re: recursion and perf as a concern. 
Regarding the example `foo.hello` and `foo.bar.hello` would only ever be counted if their explicit full name were defined as the attribute. In my testing putting in just `hello` would get no value (nothing to match). Putting in `foo.hello` will return its value but nothing else. Putting on `foo.bar.hello` returns its value and no other values. 

Or am I misunderstanding that the worry is that multiple values would be returned deeper/shallower than the defined `foo.hello` or `foo.bar.hello` explicit names? I can do more testing to verify if that is the case.

Let me know if I'm totally misunderstanding though also of course! :D
Thanks!

**Comment by djaglowski** [comment]
> Regarding the example foo.hello and foo.bar.hello would only ever be counted if their explicit full name were defined as the attribute.

I think maybe we're getting tripped up on the notation because I'm using `foo.hello` as I described [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20421#issuecomment-1490287316).

---

Hopefully more precise:

Let's use `"foo.hello"` to mean a single key that is exactly as it looks, I would not expect this key to match `"hello"` regardless of `max_depth`. 

On the other hand, if `foo["hello"]` is a key `"foo"` with a value that is an object, which contains a key `"hello"`, then this is exactly the case where I would expect we are using `max_depth`. Does that make sense?

**Comment by greatestusername** [comment]
Ah ok I get it. That makes sense to me. I appreciate the clarification! 
Will take a look at implementing that.

**Comment by greatestusername** [comment]
Changed `any_depth` to `max_depth` and added logic too the recursion that any setting of `max_depth` over 1 gets checked for nested and counts down the depth setting.

Also this now will work for any resource and I've removed the error checks/logic that was previously used to restrict to logs only.  Happy to hear your thoughts/suggestions!
Thanks!

File: connector/countconnector/counter.go [hunk 1]
```diff
@@ -48,10 +50,16 @@ type attrCounter struct {
 
 func (c *counter[K]) update(ctx context.Context, attrs pcommon.Map, tCtx K) error {
 	var errors error
+
 	for name, md := range c.metricDefs {
 		countAttrs := pcommon.NewMap()
 		for _, attr := range md.attrs {
-			if attrVal, ok := attrs.Get(attr.Key); ok {
+			attrVal, ok := attrs.Get(attr.Key)
+			if !ok && attr.MaxDepth > 1 {
```
Comment (djaglowski) [inline connector/countconnector/counter.go:58]: I think we need to tie this condition into the if/else below. Otherwise, `!ok` could result in the default value being applied.

File: connector/countconnector/counter.go [hunk 2]
```diff
@@ -117,3 +125,24 @@ func (c *counter[K]) appendMetricsTo(metricSlice pmetric.MetricSlice) {
 		}
 	}
 }
+
+func getNestedAttribute(attr string, depth int, attributes pcommon.Map) (pcommon.Value, bool) {
```
Comment (djaglowski) [inline connector/countconnector/counter.go:129]:
```suggestion
func getNestedAttribute(attr string, maxDepth int, attributes pcommon.Map) (pcommon.Value, bool) {
```
I think this is slightly more clear.

File: connector/countconnector/counter.go [hunk 3]
```diff
@@ -117,3 +125,24 @@ func (c *counter[K]) appendMetricsTo(metricSlice pmetric.MetricSlice) {
...
+	logger := zap.NewExample()
+	defer logger.Sync()
+	left, right, _ := strings.Cut(attr, ".")
+	av, ok := attributes.Get(left)
+	if ok && depth > 0 {
```
Comment (djaglowski) [inline connector/countconnector/counter.go:134]: Let's pull the depth check up to the top of the function. I think if it's 0, we can just return immediately.

File: connector/countconnector/counter.go [hunk 4]
```diff
@@ -117,3 +125,24 @@ func (c *counter[K]) appendMetricsTo(metricSlice pmetric.MetricSlice) {
 		}
 	}
 }
+
+func getNestedAttribute(attr string, depth int, attributes pcommon.Map) (pcommon.Value, bool) {
+	logger := zap.NewExample()
+	defer logger.Sync()
+	left, right, _ := strings.Cut(attr, ".")
```
Comment (djaglowski) [inline connector/countconnector/counter.go:132]: I think we're still misaligned here. We shouldn't need to split any strings.

File: connector/countconnector/counter.go [hunk 5]
```diff
@@ -117,3 +125,24 @@ func (c *counter[K]) appendMetricsTo(metricSlice pmetric.MetricSlice) {
...
+	av, ok := attributes.Get(left)
+	if ok && depth > 0 {
+		if av.Type().String() == "Slice" {
+			logger.Warn("[connector/count] cannot use Type Slice as an attribute source", zap.String("attribute", attr))
```
Comment (djaglowski) [inline connector/countconnector/counter.go:136]: I don't think we should instantiate a logger locally like this. It's not clear to me that we really need to log anything in this case anyways. Seems we can just consider it a thing we can't drill into any further, same as if we found any other data type besides a map.

File: connector/countconnector/counter.go (hunk 4 above)
Comment (greatestusername) [reply connector/countconnector/counter.go:132]: Sure. I may definitely be misunderstanding or have lost the thread. Are you saying stop cutting and doing the recursion entirely and do this a different way? Or what would you suggest doing instead?

Appreciate your thoughts and knowledge here!

**Comment by greatestusername-splunk** [comment]
committed changes to address review feedback. 

Looking for a bit more thought on what you would like to see for the nested parsing if not cutting the string and looping through. Or if the looping is what you'd like addressed? 
Happy to do whatever and primarily used the current code as it was prior art from within the repo.

Thanks in advance for your thoughts and review! 😃

File: connector/countconnector/counter.go (hunk 4 above)
Comment (djaglowski) [reply connector/countconnector/counter.go:132]: Apologies for delayed response, just getting caught up after some time away.

I think we should _not_ be cutting strings in any case, but we _should_ be recursing through nested maps.

We need a function that looks roughly like this:
```go
func countNestedAttribute(attr string, maxDepth int, attributes pcommon.Map) int {
    if maxDepth == 0 {
        return 0 // done recursing
    }
    var count int
    // Range will iterate over key/value pairs in the map.
    // The bool indicates whether or not to keep iterating.
    attributes.Range(func(k string, v pcommon.Value) bool {
        if k == attr {
            count++
            return true
        }
        if v.Type() != pcommon.ValueTypeMap {
            return true
        }
        count += countNestedAttribute(attr, maxDepth-1, v.Map())
        return true
    })
    return count
}
```

Using an example input:
```yaml
attributes:
  foo: 1
  not.foo: 0 # a different attribute, unrelated to 'foo'
  other: # a nested map that contains additional attributes
    foo: 2
```
If we call `countNestedAttribute("foo", 100, attributes)`, I'd expect the following to happen:
- We find "foo" at the top level and count +1
- We see "not.foo", but this is not the attribute we are looking for so we ignore it
- We see that "other" is a map, so we recurse by calling `countNestedAttribute("foo", 99, attributes["other"])`
  - Within the recursed call, we find "foo", so we count +1
- The end result is 2

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: Ok that makes sense and clarifies perfectly for me!
Appreciate the detailed reply and hope your time away was good!

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: After looking through this I feel like I may actually be more confused. 

The `getNestedAttribute` function in question (where the string cutting is happening) doesn't produce a count/int it produces the string value to fill in for the given attribute key. The counting of that key/value pair happens elsewhere.

Are you saying I should be doing the actual counting here or am I misunderstanding the meaning of count in this case?

Maybe the example attribute values you reference being ints is also confusing here? I'd expect them to be strings (or maybe ints as strings)?

My example data (i'm essentially trying to get out `oof` and if blank i'd give the `default_value`):
```
          - attributes:
              - key: other
                value:
                  kvlistValue: 
                    values:
                      - key: foo
                        value:
                           stringValue: oof
```

Am I totally confused here?

Comment (djaglowski) [reply connector/countconnector/counter.go:132]: I only meant to clarify _how_ attributes should be counted.

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: I guess what I'm not clear on is how that counting replaces the string cutting to get the attribute stringvalue out of the array?

It feels like we may be talking about two different things which is why I'm asking for clarification. I'm sorry if this sounds daft but I'm just really unclear on what is being asked for exactly.

Thanks! Appreciate your feedback!

**Comment by greatestusername** [comment]
been gone for a week. Will be taking another look at this this week.

File: connector/countconnector/counter.go (hunk 4 above)
Comment (djaglowski) [reply connector/countconnector/counter.go:132]: Hi Jeremy,

I don't think we can make progress on this issue until we agree that there is a difference between the following:
1) an attribute whose _value_ is an map that contains other attributes
2) an attribute whose _key_ contains one or more `.`

```
attributes:
  hello:
    foo: oof
  hello.foo: oof
```

---

If we are trying to count attributes called "foo", then only the first case above should be counted.

String splitting is only necessary if you are trying to count "hello.foo" as "foo", which is incorrect.

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: That approach doesn't actually provide an answer though as it isn't returning the value from the `kvlist`. It simply returns a count of how many times the recursion in the function happened which doesn't help with getting the value string out of the nested key.

Or am I totally missing something here?

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: In the mean time I'm doing this in the transform processor to be able to count nested fields
`- set(attributes["nested.attribute.foo"], attributes["nested"]["attribute"]["foo"]) where attributes["nested"]["attribute"]["foo"] != nil`

Which turns the nested key and value into a single named dot delimited key and associated value that isn't buried in a map/`kvlist`

Comment (djaglowski) [reply connector/countconnector/counter.go:132]: @greatestusername, if we agree that `attributes["nested.attribute.foo"]` is different than `attributes["nested"]["attribute"]["foo"]`, then I don't understand why the solution would involve tokenizing `"nested.attribute.foo"`. What am I missing here?

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: I think the primary confusion is around not being able to reference `attributes["nested"]["attribute"]["foo"]` directly in the config and having to use the "made up" string of `nested.attribute.foo`?

If I could reference it as `attributes["nested"]["attribute"]["foo"]` in the config rather than `nested.attribute.foo` there wouldn't be an issue as we'd be able to reference the nested maps explicitly. 

```
  count:
    logs:
      logs.to.metric:
        description: turning logs into metrics
        attributes:
          - key: attributes["nested"]["attribute"]["foo"]
            default_value: None 
```
Would then describe we want the value of `foo` under `attribute` map which is under `nested` map. 

But as it stands the only config I can make stick is:
```        
  count:
    logs:
      logs.to.metric:
        description: turning logs into metrics
        attributes:
          - key: nested.attribute.foo
            default_value: None 
```
Which I think is causing a lot of the confusion in this thread because `nested.attribute.foo` is "made up" and not an actual OTTL description of the data path?

In other words would it make more sense to have the `- key:` config accept standard OTTL syntax?

Comment (djaglowski) [reply connector/countconnector/counter.go:132]: @greatestusername, thanks for detailing this out. I agree with where you're going here. Let's use OTTL syntax for referring to nested attributes.

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: Would this be a larger change to the connector than initially proposed? I.E. would this change the config entirely to always expect OTTL? in the `key` part of the config? Regardless of if it was a nested or non-nested attribute?

Before doing this work I'd want to outline with you your expectations and what you'd like to see as it is a larger set of changes I'd imagine than the issue initially scoped.

What are your thoughts on this?

Apologies for the couple days without a response. Life took over.
Thank you again for your help, review, and patience!

Comment (djaglowski) [reply connector/countconnector/counter.go:132]: @greatestusername, it's not immediately clear to me what the solution should be. Perhaps it would would be helpful to include an example config that shows a possible solution.

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: This config from my example above would theoretically be what I'd expect for the example data I've been discussing if `key` accepted and used `OTTL` [syntax/conventions](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/README.md#paths).
```
  count:
    logs:
      logs.to.metric:
        description: turning nested attribute into metric
        attributes:
          - key: attributes["nested"]["attribute"]["foo"]
            default_value: None 
```

Essentially that is just using the OTTL conventions in the `key` section of the config.

That is what brings up my previous question though. Should `key` be changed to always be expecting OTTL syntax/conventions generally?

In that case getting an attribute like `single.dot.named.key` would be configured like 
```
  count:
    logs:
      logs.to.metric:
        description: turning attribute into count
        attributes:
          - key: attributes["single.dot.named.key"]
            default_value: None 
```
Where currently the config for `key` would just be `- key: single.dot.named.key` (not using OTTL syntax).

In other words, should this be a larger change to the connector (using OTTL syntax for all `key` config attributes)?

Appreciate your continued thoughts and review on this!

Comment (djaglowski) [reply connector/countconnector/counter.go:132]: Thanks for expanding this for me.

I see why it would require separate formats, but I think the `attributes["..."]` framing would seem extraneous to most users who only care about non-nested attributes.

My opinion is that it would be reasonable to support both formats for now. Essentially by adding the `attributes["..."]` wrapper automatically to `single.do.named.key`. WDYT?

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: So the "backend" would always be using OTTL syntax ("essentially by adding the `attributes["..."]` wrapper automatically")?

I.E. The connector would do a check on `key` from the config and if it didnt have `attributes[]` syntax it would add that before doing the OTTL evaluation on any `key` provided?

Thus `-key: single.dot.named.key` would be changed to `attributes["single.dot.named.key"]` inside of the connector internal functions. 
But `-key: attributes["single.dot.named.key"]` would not be changed because it is already in the OTTL syntax and get used as written.


Just want to verify that I'm understanding what you're saying the way you're saying it.

Comment (djaglowski) [reply connector/countconnector/counter.go:132]: I think that makes sense. WDYT?

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: That seems reasonable to me as a solution to minimize any breakage to current users.

Comment (greatestusername) [reply connector/countconnector/counter.go:132]: Just wanted to confirm this is the path forward you'd like to see before starting in on it. We good? 😃

Comment (djaglowski) [reply connector/countconnector/counter.go:132]: Yes, I believe we're on the same page.
